src/gui/dialogs/statistics_dialog.cpp

```diff
@@ -133,34 +151,169 @@ void statistics_dialog::add_damage_row(
+	std::string hitrate_str;
+	// The string with the a priori probability of that result
+	std::string percentage_str;
+};
+
+// Return the strings to use in the "Hits" table, showing actual and expected number of hits.
+static hitrate_table_element tally(const statistics::stats::hitrate_map& by_cth, const bool more_is_better)
+{
+	unsigned int overall_hits = 0;
+	double expected_hits = 0;
+	unsigned int overall_strikes = 0;
+
+	for(const auto& i : by_cth) {
+		int cth = i.first;
+		overall_hits += i.second.hits;
+		expected_hits += (cth * 0.01) * i.second.strikes;
+		overall_strikes += i.second.strikes;
+	}
+
+	std::ostringstream str, str2;
+	str << overall_hits << " / " << expected_hits;
+
+	// Compute the a priori probability of this actual result, by simulating many attacks against a single defender.
+	{
+		config defender_cfg(
+			"id", "statistics_dialog_dummy_defender",
+			"hide_help", true,
+			"do_not_list", true,
+			"hitpoints", overall_strikes
+		);
+		unit_type defender_type(defender_cfg);
+		unit_types.build_unit_type(defender_type, unit_type::BUILD_STATUS::FULL);
+
+		battle_context_unit_stats defender_bc(&defender_type, nullptr, false, nullptr, nullptr, 0 /* not used */);
+		std::unique_ptr<combatant> current_defender(new combatant(defender_bc));
+
+		for(const auto& i : by_cth) {
+			int cth = i.first;
+			config attacker_cfg(
+				"id", "statistics_dialog_dummy_attacker" + std::to_string(cth),
+				"hide_help", true,
+				"do_not_list", true,
+				"hitpoints", 1
+			);
+			unit_type attacker_type(attacker_cfg);
+			unit_types.build_unit_type(attacker_type, unit_type::BUILD_STATUS::FULL);
+
+			attack_ptr attack = std::make_shared<attack_type>(config(
+				"type", "blade",
+				"range", "melee",
+				"name", "dummy attack",
+				"damage", 1,
+				"number", i.second.strikes
+			));
+
+			battle_context_unit_stats attacker_bc(&attacker_type, attack, true, &defender_type, nullptr, 100 - cth);
+			defender_bc = battle_context_unit_stats(&defender_type, nullptr, false, &attacker_type, attack, 0 /* not used */);
+
+			// Update current_defender with the new defender_bc.
+			current_defender.reset(new combatant(*current_defender, defender_bc));
+
+			combatant attacker(attacker_bc);
+			attacker.fight(*current_defender);
+		}
+
+		const std::vector<double>& final_hp_dist = current_defender->hp_dist;
+		const auto& chance_of_exactly_N_hits = [&final_hp_dist](int n) { return final_hp_dist[final_hp_dist.size() - 1 - n]; };
+
+		double probability_lt = 0.0;
+		for(unsigned int i = 0; i < overall_hits; ++i) {
+			probability_lt += chance_of_exactly_N_hits(i);
+		}
+		double probability_gt = 0.0;
+		for(unsigned int i = final_hp_dist.size() - 1; i > overall_hits; --i) {
+			probability_gt += chance_of_exactly_N_hits(i);
+		}
+
+		if(overall_strikes == 0) {
+			// Start of turn
+			str2 << "0%";
+		} else {
+			const auto& add_probability = [&str2](double probability, bool more_is_better) {
```

Comment:
Should not be a reference.

Reply:
21dc2351a2751ff0ac3c91f558e0450a429c0bfa